code/Player/PropPossession.cs

```diff
@@ -0,0 +1,207 @@
+
+	protected override void OnActivate()
+	{
+		_player = Entity;
+
+		if ( Host.IsServer )
+		{
+			_player.Camera = new FollowEntityCamera( Prop );
+			Prop.Owner = Entity;
+			_timeUntilRecharge = Game.PropPossessionRechargeTime;
+		}
+		else
+		{
+			if ( Local.Pawn is Player player && !player.IsAlive() )
+				_nameplate = new PossessionNameplate( Entity, Prop );
+
+			if ( Entity.IsLocalPawn )
+				_info = new PossessionInfo( this ); // local pawn is possessing a prop
+		}
+	}
+
+	protected override void OnDeactivate() // `Entity` property is null when this is called
+	{
+		if ( Host.IsServer )
+		{
+			if ( _player.IsValid() && !_player.IsAlive() )
+				_player.Camera = new FreeSpectateCamera();
+
+			if ( Prop.IsValid() )
+				Prop.Owner = null;
+		}
+		else
+		{
+			_nameplate?.Delete();
+			_info?.Delete();
+		}
+
+		_player = null;
+	}
+
+	[GameEvent.Player.StatusChanged]
+	private void OnStatusChanged( Player somePlayer, PlayerStatus oldStatus )
+	{
+		if ( somePlayer.Status == PlayerStatus.Alive )
+		{
+			// somePlayer has come alive
+			if ( Host.IsServer && somePlayer == Entity )
+			{
+				// _player is not possessing anything anymore since they have come alive
+				// we cannot immediately remove this component since that would remove all event handlers
+				// (including the one we are currently in) which would cause a InvalidOperationException
+				// (Collection was modified; enumeration operation may not execute)
+				RemoveThisComponentDelayed();
+			}
+
+			if ( Host.IsClient && somePlayer.IsLocalPawn )
+				_nameplate?.Delete(); // local player has come alive, must loose all nameplates
+		}
+		else if ( somePlayer.Status != PlayerStatus.Alive )
+		{
+			// somePlayer has died
+			if ( somePlayer.IsLocalPawn )
+				_nameplate = new PossessionNameplate( _player, Prop ); // local player has died => show nameplates
+		}
+	}
+
+	private async void RemoveThisComponentDelayed()
+	{
+		await GameTask.Delay( 1 );
+		_player?.Components.RemoveAny<PropPossession>();
+	}
+
+	[Event.Tick.Server]
+	private void RechargePunchesAndCheckProp()
+	{
+		if ( !Prop.IsValid() )
+		{
+			RemoveThisComponentDelayed(); // delayed removal for the same reason as above
+			return;
+		}
+
+		if ( _timeUntilRecharge )
+		{
+			Punches = Math.Min( Punches + 1, Game.PropPossessionMaxPunches );
+			_timeUntilRecharge = Game.PropPossessionRechargeTime;
+		}
+	}
+
+	[Event.BuildInput]
+	private void BuildInput( InputBuilder input )
+	{
+		if ( !Entity.IsValid() || !Entity.IsLocalPawn )
+			return;
+
+		if ( input.Pressed( InputButton.Duck ) )
+		{
+			CancelPossession();
+			return;
+		}
+
+		float forward = input.Pressed( InputButton.Forward ) ? 1f : (input.Pressed( InputButton.Back ) ? -1f : 0f);
+		float left = input.Pressed( InputButton.Left ) ? 1f : (input.Pressed( InputButton.Right ) ? 1f : 0f);
+		bool up = input.Pressed( InputButton.Jump );
+		var rotation = Rotation.From( input.ViewAngles );
+
+		if ( forward + left != 0f || up )
```

Comment:
Is there a different way to handle this? I'll spend some time thinking about it.